### ISSUE TITLE
Missing ) in SQL query using top_n() on Spark 1.6.x

### ISSUE THREAD

**Issue by richierocks** [rstudio/sparklyr#645]
To reproduce:

```r
library(dplyr)
library(sparklyr)

sc <- spark_connect("local")

mtcars_tbl <- copy_to(sc, mtcars)

mtcars_tbl %>% 
  count(mpg) %>% 
  top_n(5)
## Error: org.apache.spark.sql.AnalysisException: line 2:59 missing ) at ',' near ')' in subquery source
##   line 3:0 cannot recognize input near 'AS' 'zzz2' 'FROM' in subquery source;
```

**Comment by richierocks** [comment]
Actually, there seem to be several problems with variations on this code.

For example, passing `n` as the weight argument fails, presumably because it thinks the column `n` is actually the dplyr function `n()`.
 
```r
mtcars_tbl %>% 
  count(mpg) %>% 
  top_n(5, wt = n)
## Error: Unknown input type: function
```

This uncovers a problem with renaming columns.

```r
mtcars_tbl %>% 
  count(mpg) %>% 
  rename(nn = n) 
## Source:     lazy query [?? x 2]
## Database:   spark_connection
## Error: `x` must be a vector
```

It isn't clear to me if these are separate issues, or one issue with the SQL count statement that is generated.

**Comment by richierocks** [comment]
Oh, sorry, the minimal reproducible example is actually even simpler.

```r
mtcars_tbl %>%
  top_n(5)
## Error: org.apache.spark.sql.AnalysisException: line 2:117 missing ) at ',' near ')' in subquery source
##   line 3:0 cannot recognize input near 'AS' 'zzz23' 'FROM' in subquery source;
```

**Comment by kevinykuo** [comment]
@richierocks I can't reproduce -- could you try updating/reloading dplyr and dbplyr?

Also does this return anything for you?

```
> mtcars_tbl %>%
+   top_n(5) %>%
+   explain()
Selecting by carb
<SQL>
SELECT `mpg`, `cyl`, `disp`, `hp`, `drat`, `wt`, `qsec`, `vs`, `am`, `gear`, `carb`
FROM (SELECT `mpg`, `cyl`, `disp`, `hp`, `drat`, `wt`, `qsec`, `vs`, `am`, `gear`, `carb`, rank() OVER (ORDER BY `carb` DESC) AS `zzz8`
FROM `mtcars`) `sgdhsaqujg`
WHERE (`zzz8` <= 5.0)


<PLAN>
```

**Comment by richierocks** [comment]
`explain()` suggests that the `ORDER BY` statement is the problem.

```r
mtcars_tbl %>%
+   top_n(5) %>%
+   explain()
Selecting by carb
<SQL>
SELECT `mpg`, `cyl`, `disp`, `hp`, `drat`, `wt`, `qsec`, `vs`, `am`, `gear`, `carb`
FROM (SELECT `mpg`, `cyl`, `disp`, `hp`, `drat`, `wt`, `qsec`, `vs`, `am`, `gear`, `carb`, rank() OVER (ORDER BY (4.0, 4.0, 1.0, 1.0, 2.0, 1.0, 4.0, 2.0, 2.0, 4.0, 4.0, 3.0, 3.0, 3.0, 4.0, 4.0, 4.0, 1.0, 2.0, 1.0, 1.0, 2.0, 2.0, 4.0, 2.0, 1.0, 2.0, 2.0, 4.0, 6.0, 8.0, 2.0) DESC) AS `zzz2`
FROM `mtcars`) `iivwexouyo`
WHERE (`zzz2` <= 5.0)
```

I built all the packages today.

```r
c("sparklyr", "dplyr", "dbplyr", "rlang") %>% 
  setNames(., .) %>% 
  lapply(packageDescription, fields = c("Version", "Built", "GithubSHA1"))
```

```
$sparklyr
Version: 0.5.3-9005
Built: R 3.3.3; ; 2017-04-28 14:26:54 UTC; unix
GithubSHA1: 4794dd7120f7bee437007d898319424a11e59e8d

-- File: /Library/Frameworks/R.framework/Versions/3.3/Resources/library/sparklyr/Meta/package.rds 
-- Fields read: Version, Built, GithubSHA1

$dplyr
Version: 0.5.0.9004
Built: R 3.3.3; x86_64-apple-darwin13.4.0; 2017-04-28 14:23:45 UTC; unix
GithubSHA1: e0c866dfbb25124e770b34dd4483b95a7677519f

-- File: /Library/Frameworks/R.framework/Versions/3.3/Resources/library/dplyr/Meta/package.rds 
-- Fields read: Version, Built, GithubSHA1

$dbplyr
Version: 0.0.0.9001
Built: R 3.3.3; ; 2017-04-28 14:26:52 UTC; unix
GithubSHA1: c4bd1be9677032736de1b5b857cd215a74279416

-- File: /Library/Frameworks/R.framework/Versions/3.3/Resources/library/dbplyr/Meta/package.rds 
-- Fields read: Version, Built, GithubSHA1

$rlang
Version: 0.0.0.9018
Built: R 3.3.3; x86_64-apple-darwin13.4.0; 2017-04-28 14:23:39 UTC; unix
GithubSHA1: b22ce1ce7cdb951933f26b30142f57bf9b88419d

-- File: /Library/Frameworks/R.framework/Versions/3.3/Resources/library/rlang/Meta/package.rds 
-- Fields read: Version, Built, GithubSHA1
```

Will continue to investigate.

**Comment by richierocks** [comment]
Further investigation reveals that the problems occur with Spark `1.6.2`, but not Spark `2.0.2`.

**Comment by kevinushey** [comment]
Here's the generated query with Spark 1.6.2:

```
> sql_render(query)
<SQL> SELECT `mpg`, `cyl`, `disp`, `hp`, `drat`, `wt`, `qsec`, `vs`, `am`, `gear`, `carb`
FROM (SELECT `mpg`, `cyl`, `disp`, `hp`, `drat`, `wt`, `qsec`, `vs`, `am`, `gear`, `carb`, rank() OVER (ORDER BY (4.0, 4.0, 1.0, 1.0, 2.0, 1.0, 4.0, 2.0, 2.0, 4.0, 4.0, 3.0, 3.0, 3.0, 4.0, 4.0, 4.0, 1.0, 2.0, 1.0, 1.0, 2.0, 2.0, 4.0, 2.0, 1.0, 2.0, 2.0, 4.0, 6.0, 8.0, 2.0) DESC) AS `zzz6`
FROM `mtcars`) `ofukbymqme`
WHERE (`zzz6` <= 5.0)
```

The query looks fine and is unchanged with Spark 2.1.0, so I assume this is a bug in Spark 1.6.2.

**Comment by duttashi** [comment]
@kevinushey  @richierocks  @javierluraschi @kevinykuo , I tried reproducing this issue, and I got no errors. It worked fine. 
So, I doubt that it is a bug in Spark 1.6.2 

My session information is as follows;

```
> sessionInfo()
R version 3.3.3 (2017-03-06)
Platform: i386-w64-mingw32/i386 (32-bit)
Running under: Windows 7 x64 (build 7601) Service Pack 1

locale:
[1] LC_COLLATE=English_United States.1252  LC_CTYPE=English_United States.1252   
[3] LC_MONETARY=English_United States.1252 LC_NUMERIC=C                          
[5] LC_TIME=English_United States.1252    

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] DBI_0.7            sparklyr_0.5.6     magrittr_1.5       dplyr_0.7.0        nycflights13_0.2.2

loaded via a namespace (and not attached):
 [1] Rcpp_0.12.11     rstudioapi_0.6   hms_0.3          rappdirs_0.3.1   xtable_1.8-2    
 [6] R6_2.2.2         rlang_0.1.1      httr_1.2.1       tools_3.3.3      parallel_3.3.3  
[11] config_0.2       dbplyr_1.0.0     withr_1.0.2      htmltools_0.3.6  yaml_2.1.14     
[16] assertthat_0.2.0 rprojroot_1.2    digest_0.6.12    tibble_1.3.3     shiny_1.0.3     
[21] readr_1.1.1      base64enc_0.1-3  glue_1.1.1       mime_0.5         backports_1.1.0 
[26] jsonlite_1.5     httpuv_1.3.3 
```
#### load libraries
```
>library(sparklyr)
>library(dplyr)
```

#### Copy data to spark local instance
`>copy_to(sc, mtcars)`
#### check data tables in spark local cluster
```
>src_tbls(sc)
[1] "mtcars"
```
#### print the spark version
```
>spark_version(sc)
[1] ‘1.6.2’
```
#### execute dplyr commands
```
> mtcars %>%
+   count(mpg)%>%
+   top_n(5)
Selecting by  n
# A tibble: 7 x 2
    mpg     n
  <dbl> <int>
1  10.4     2
2  15.2     2
3  19.2     2
4  21.0     2
5  21.4     2
6  22.8     2
7  30.4     2

> mtcars %>%
+   count(mpg)%>%
+   top_n(5, wt=n)
# A tibble: 7 x 2
    mpg     n
  <dbl> <int>
1  10.4     2
2  15.2     2
3  19.2     2
4  21.0     2
5  21.4     2
6  22.8     2
7  30.4     2

> mtcars %>%
+   count(mpg)%>%
+   rename(nn=n)
# A tibble: 25 x 2
     mpg    nn
   <dbl> <int>
 1  10.4     2
 2  13.3     1
 3  14.3     1
 4  14.7     1
 5  15.0     1
 6  15.2     2
 7  15.5     1
 8  15.8     1
 9  16.4     1
10  17.3     1
# ... with 15 more rows
```

**Comment by javierluraschi** [comment]
Right, this one I remember was a `dplyr` regression which is fixed now in `dplyr 0.7`. Closing.
